engine/class_modules/warlock/sc_warlock_affliction.cpp

```diff
@@ -106,10 +106,28 @@ struct shadow_bolt_t : public affliction_spell_t
   void impact( action_state_t* s ) override
   {
     affliction_spell_t::impact( s );
-    if ( result_is_hit( s->result ) && ( !p()->min_version_check( VERSION_9_1_0 ) || p()->talents.shadow_embrace->ok() ) )
+    if ( result_is_hit( s->result ) )
     {
-      // Add passive check
-      td( s->target )->debuffs_shadow_embrace->trigger();
+      if ( !p()->min_version_check( VERSION_9_1_0 ) || p()->talents.shadow_embrace->ok() )
+      {
+        // Add passive check
+        td( s->target )->debuffs_shadow_embrace->trigger();
+      }
+
+      // TODO: Check PTR to see if 4-set procs on cast finish, or on damage
+      if ( p()->sets->has_set_bonus( WARLOCK_AFFLICTION, T28, B4 ) )
+      {        
+        auto tdata = this->td( s->target );
+        bool tierDotsActive = tdata->dots_agony->is_ticking();
```

Comment:
I would express this with a single boolean statement instead of (mis)using the bitwise operator.

Reply:
Would this be functionally different from the existing implementation in any way?

Reply:
Problem here isn't functionality but meaning. `&` is a bitwise operation (and so is `&=`) but you want a logical operation. Sadly C++ doesn't have logical version of `&=` such as `&&=`, so you're left with either `val = A && B;` or `val = A; val = val && B;`.

By the way, those two are not the same when dealing with expression that can be converted to `bool`:
```cpp
bool b = true;
b &= 2;
// b is false

bool c = true;
c = c && 2;
// c is true
```

Reply:
I am aware of the bitwise nature of the assignment operators here but did not see how it was functionally different in the event that we are working purely with boolean data types. Since all of these operators are boolean in nature ( is_ticking() ) and would have no reason to become non-boolean values in the future, this interaction should not possibly arise. The only difference in functionality is the short-circuit nature being lost in the expression.

Reply:
It's the same reason we don't use `+` or `*` on booleans (even though it technically works as disjunction/conjuction when all the arguments are booleans): it's just the wrong operator to use.